### PR TITLE
responsive compiler prototype: parser to produce new AST classes

### PR DESCRIPTION
**Pull request by mppf**
This PR contains a separate parser for the compiler revamp effort.

Split out from PR #17583, which is a start at the compiler revamp effort
described in the
[1.24 release notes ongoing efforts](https://chapel-lang.org/releaseNotes/1.24/04-ongoing.pdf).

The flex/bison files started out as copies of the parser from the main
compiler. I removed all semantic actions and adjusted these to not use
global variables. Additionally, instead of storing comments in bison
locations, it stores a vector of comments not yet consumed in the
ParserContext structure. Along with removing the semantic actions I have
adjusted the types of these rules and made a couple of minor changes to
get the comments processing correctly. As far as I know, this parser no
longer uses any global state. Several rules that used to return a
BlockStmt (because they are a "statement) now return CommentsAndStmt. The
idea here is that, during parsing, when we create a statement, we gather
comments that appeared before it. We include those in the result of
parsing that statement. Additionally there is some logic to clear the
comments when we reach the end of a statement.

While working on the parser, I moved lots of the C++ support code that
used to be in chapel.flex/chapel.bison to other files that are
`#included` in these. That way, we don't have to regenerate the parser if
they change.

Note that in the lexer, I've changed the handling of comments to include
recording the entire comment (rather than leaving out the introduction or
dealing with other formatting). That way, tools like `chpldoc` can
separately decide which comment introductions require further processing.

Additionally, in the lexer, I've changed the parsing of string literals
to record string data in the unescaped form. Previously, the compiler
stored string literals with C escapes. Storing the escapes has led to a
variety of bugs over time (because things like string.length for param
strings needs to handle the escapes). But more importantly, as the
default backend switches to LLVM, these C escapes don't make sense
anymore in the default configuration. Instead, we will need to add the
necessary escapes when generating C code.

The parser in this PR has the ability to parse a simple subset of the
language (including `{ }` blocks, trivial variable declarations like `var
a;`, and identifier statements like `a;`). Other language elements still
have parser rules but their semantic actions just create error sentinel
nodes.

The parser currently stores the results from the rules in a union. As a
result, we can't have types that have default constructors. (This is the
reason for the `PODUniqueString` type and also a reason that the parser
code has its own kind of expression list type, etc). Bison has some
support for C++ variants but I did not take on the task of updating our
parser to use those. As it is today, the nested rules from a parser rule
generally return a pointer that needs to be stored somewhere or deleted.
This is manageable, although it makes the parser code less elegant than
it could be.

Lastly, I created several helper `#define`s that can be used in the
semantic actions. This was important because for now most of the semantic
actions are "TODO" and so I needed a short way of putting that. But,
these macros also help with things like handling comments appropriately
when parsing a Statement. These macros like `TODOSTMT` `LOC` `STR` etc
are defined in parser-help.h.

 - [x] full local testing

Reviewed by @lydia-duncan - thanks!

### CODE REVIEW

**Review by lydia-duncan — APPROVED** [review]
The only essential change is the typo fix, the rest I'm okay with you ignoring if you don't feel up for it :)

I believe this is saving more comments than we would be in the previous version of the compiler, but I think that makes sense in the new world and would rather have too many comments saved (as long as I can tell where each ends) than miss them.

File: compiler/next/lib/Frontend/Parser/chapel.lex [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+/*
+   Provide a condition stack
+   This is used to alter the handling of "{" when it appears immediately after "extern"
+*/
+
+%option stack
+
+%{
+
+#include "bison-chapel.h"
```
[the remark this answers is not in the document]
Comment (mppf) [reply compiler/next/lib/Frontend/Parser/chapel.lex:50]: This one is not a public interface (in the sense of the compiler-as-a-library) and so it's in the lib/... subdirectory and we just include it from the current directory

File: compiler/next/lib/Frontend/Parser/chapel.lex [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+"\"\"\""         return processTripleStringLiteral(yyscanner, "\"", STRINGLITERAL);
+"'''"            return processTripleStringLiteral(yyscanner, "'", STRINGLITERAL);
+"b\"\"\""        return processTripleStringLiteral(yyscanner, "\"", BYTESLITERAL);
+"b'''"           return processTripleStringLiteral(yyscanner, "'", BYTESLITERAL);
+"\""             return processStringLiteral(yyscanner, "\"", STRINGLITERAL);
+"\'"             return processStringLiteral(yyscanner, "\'", STRINGLITERAL);
+"b\""            return processStringLiteral(yyscanner, "\"", BYTESLITERAL);
+"b\'"            return processStringLiteral(yyscanner, "\'", BYTESLITERAL);
+"c\""            return processStringLiteral(yyscanner, "\"", CSTRINGLITERAL);
+"c\'"            return processStringLiteral(yyscanner, "\'", CSTRINGLITERAL);
```
[the remark this answers is not in the document]
Comment (mppf) [reply compiler/next/lib/Frontend/Parser/chapel.lex:314]: I can take out the escaping of `'` on 312 and 310

File: compiler/next/lib/Frontend/Parser/chapel.ypp [hunk 3]
```diff
@@ -0,0 +1,2510 @@
...
+//
+// Chapel Parser Conventions
+//
+// opt_       -- optional
+//      _expr -- expression
+//      _ls   -- list
+//      _stmt -- statement
+//      _type -- type
+//
+// The rules are listed in depth-first order from stmt and expr
```
[the remark this answers is not in the document]
Comment (mppf) [reply compiler/next/lib/Frontend/Parser/chapel.ypp:30]: I don't think so; the parser still has an idea of statement. It's just not present in the AST it builds.

File: compiler/next/lib/Frontend/Parser/chapel.ypp [hunk 4]
```diff
@@ -0,0 +1,2510 @@
...
+    PrototypeTag_DEFAULT,
+    PrototypeTag_PROTOTYPE,
+  };
+
+  enum ThrowsTag {
+    ThrowsTag_DEFAULT,
+    ThrowsTag_THROWS,
+  };
+
+  enum NewTag {
```
[the remark this answers is not in the document]
Comment (mppf) [reply compiler/next/lib/Frontend/Parser/chapel.ypp:107]: I'm going to move it to an AST class once we define one for `new`. So I'll just put a comment to that effect for now.

File: compiler/next/lib/Frontend/Parser/chapel.ypp [hunk 5]
```diff
@@ -0,0 +1,2510 @@
...
+    IntentTag_OUT,
+    IntentTag_INOUT,
+    IntentTag_PARAM,
+    IntentTag_TYPE,
+  };
+
+  enum LinkageTag {
+    LinkageTag_DEFAULT,
+    LinkageTag_INLINE,
+    LinkageTag_OVERRIDE,
```
[the remark this answers is not in the document]
Comment (mppf) [reply compiler/next/lib/Frontend/Parser/chapel.ypp:149]: Like the NewTag enum, this one will get pulled out into an AST class and we can discuss it further then. Even if we want some combinations we can have e.g. LinkageTag_INLINE_EXPORT. I can't think of any combinations that make sense today, though.

File: compiler/next/lib/Frontend/Parser/chapel.ypp [hunk 6]
```diff
@@ -0,0 +1,2510 @@
...
+    PotentialRename(owned<Expr> elem) : PotentialRename(elem.release()) { }
+    PotentialRename(Expr* from, Expr* to)
+    {
+      this->tag = DOUBLE;
+      this->elem = nullptr;
+      this->renamed = std::make_pair(from, to);
+    }
+    PotentialRename(owned<Expr> from, owned<Expr> to)
+    : PotentialRename(from.release(), to.release()) { }
+    PotentialRename(Expr* from, owned<Expr> to)
```
[the remark this answers is not in the document]
Comment (mppf) [reply compiler/next/lib/Frontend/Parser/chapel.ypp:177]: If we need that in a rule/semantic action, we will add it at that point. IIRC the others here were used somewhere.

File: compiler/next/lib/Frontend/Parser/chapel.ypp [hunk 7]
```diff
@@ -0,0 +1,2510 @@
...
+
+
+  // These currently all need to be POD types (i.e. pointers and ints)
+  // rather than say std::vector to keep bison working with the current
+  // strategy. In the future we could probably switch to a more
+  // C++ mode of using bison.
+  union YYSTYPE {
+    // The lexer only uses these two
+    PODUniqueString            uniqueStr;
+    SizedStr                   sizedStr;
```
[the remark this answers is not in the document]
Comment (mppf) [reply compiler/next/lib/Frontend/Parser/chapel.ypp:236]: Are you saying you want

```
    PODUniqueString uniqueStr;
    SizedStr sizedStr;
```

etc?

File: compiler/next/lib/Frontend/Parser/chapel.ypp [hunk 8]
```diff
@@ -0,0 +1,2510 @@
...
+;
+
+req_formal_ls:
+  TLP formal_ls TRP  { $$ = $2; }
+;
+
+formal_ls:
+                           { $$ = TODOLIST(@$); }
+| formal                   { $$ = TODOLIST(@$); }
+| formal_ls TCOMMA formal  { $$ = TODOLIST(@$); }
```
[the remark this answers is not in the document]
Comment (mppf) [reply compiler/next/lib/Frontend/Parser/chapel.ypp:1696]: Indeed, this compiles with the regular compiler

``` chapel
proc foo(, x: int) { }

foo(1);
```

File: compiler/next/lib/Frontend/Parser/chapel.ypp [hunk 9]
```diff
@@ -0,0 +1,2510 @@
...
+| TLP TDOTDOTDOT expr TRP
+    { $$ = TODOEXPR(@$); }
+| expr TCOLON expr
+    { $$ = TODOEXPR(@$); }
+| expr TDOTDOT expr
+    { $$ = TODOEXPR(@$); }
+| expr TDOTDOTOPENHIGH expr
+    { $$ = TODOEXPR(@$); }
+
+/* The following cases would extend the current '..<' open range
```
[the remark this answers is not in the document]
Comment (mppf) [reply compiler/next/lib/Frontend/Parser/chapel.ypp:2335]: Sure, I can move this comment just before the `;`.

File: compiler/next/lib/Frontend/Parser/chapel.ypp (hunk 8 above)
[the remark this answers is not in the document]
Comment (mppf) [reply compiler/next/lib/Frontend/Parser/chapel.ypp:1696]: I'll try to fix it.

[the remark this answers is not in the document]
Comment (mppf) [reply compiler/next/lib/Frontend/Parser/chapel.ypp]: #17607 will fix it in the regular compiler and I've made a similar adjustment in the new parser.

File: compiler/next/lib/Frontend/Parser/chapel.ypp (hunk 7 above)
[the remark this answers is not in the document]
Comment (lydia-duncan) [reply compiler/next/lib/Frontend/Parser/chapel.ypp:236]: Yup!

File: compiler/next/lib/Frontend/Parser/chapel.ypp (hunk 8 above)
[the remark this answers is not in the document]
Comment (lydia-duncan) [reply compiler/next/lib/Frontend/Parser/chapel.ypp]: Awesome, thanks!

File: compiler/next/lib/Frontend/Parser/chapel.ypp (hunk 7 above)
[the remark this answers is not in the document]
Comment (mppf) [reply compiler/next/lib/Frontend/Parser/chapel.ypp:236]: I did not get this into this PR but #17609 will get to it.

**Pull request merged**